crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -0,0 +1,362 @@
+    pub(super) fn unbound() -> Self {
+        Self {
+            visible_definitions: Definitions::default(),
+            constraints: Constraints::default(),
+            may_be_unbound: true,
+        }
+    }
+
+    /// Return a new [`SymbolState`] representing a symbol with a single visible definition.
+    pub(super) fn with(definition_id: ScopedDefinitionId) -> Self {
+        Self {
+            visible_definitions: Definitions::with(definition_id.into()),
+            constraints: Constraints::of_size(1),
+            may_be_unbound: false,
+        }
+    }
+
+    /// Add Unbound as a possibility for this symbol.
+    pub(super) fn add_unbound(&mut self) {
+        self.may_be_unbound = true;
+    }
+
+    /// Add given constraint to all currently-visible definitions.
+    pub(super) fn add_constraint(&mut self, constraint_id: ScopedConstraintId) {
+        self.constraints.insert_in_each(constraint_id.into());
+    }
+
+    /// Merge two [`SymbolState`] and return the result.
+    pub(super) fn merge(a: &SymbolState, b: &SymbolState) -> SymbolState {
+        let mut merged = Self {
+            visible_definitions: Definitions::default(),
+            constraints: Constraints::default(),
+            may_be_unbound: a.may_be_unbound || b.may_be_unbound,
+        };
+        let mut a_defs_iter = a.visible_definitions.iter();
+        let mut b_defs_iter = b.visible_definitions.iter();
+        let mut a_constraints_iter = a.constraints.iter();
+        let mut b_constraints_iter = b.constraints.iter();
+
+        let mut opt_a_def: Option<u32> = a_defs_iter.next();
+        let mut opt_b_def: Option<u32> = b_defs_iter.next();
+
+        // Iterate through the definitions from `a` and `b`, always processing the lower definition
+        // ID first, and pushing each definition onto the merged `SymbolState` with its
+        // constraints. If a definition is found in both `a` and `b`, push it with the intersection
+        // of the constraints from the two paths; a constraint that applies from only one possible
+        // path is irrelevant.
+
+        // Helper to push `def`, with constraints in `constraints_iter`, onto  merged`.
+        let push = |def, constraints_iter: &mut ConstraintsIterator, merged: &mut Self| {
+            merged.visible_definitions.insert(def);
+            let Some(constraints) = constraints_iter.next() else {
+                // SAFETY: we only ever create SymbolState with either no definitions and no
+                // constraint bitsets (`::unbound`) or one definition and one constraint bitset
+                // (`::with`), and `::merge` always pushes one definition and one constraint bitset
+                // together (just below), so the number of definitions and the number of constraint
+                // bitsets can never get out of sync.
+                unreachable!("definitions and constraints length mismatch");
+            };
```

Comment:
Nit: I would use `expect` here instead of unreachable.
```suggestion
            // SAFETY: we only ever create SymbolState with either no definitions and no
            // constraint bitsets (`::unbound`) or one definition and one constraint bitset
            // (`::with`), and `::merge` always pushes one definition and one constraint bitset
            // together (just below), so the number of definitions and the number of constraint
            // bitsets can never get out of sync.
            let constraints = constraints_iter.next().expect("definitions and constraints length mismatch");
```